python/GafferUI/VectorDataWidget.py

```diff
@@ -1338,6 +1381,113 @@ def _createEditorInternal( self, index ) :
+
+		self.__colorChooser = None
+		self.__popup = None
+
+	def paint( self, painter, option, index ) :
+
+		isColor, value = self.__colorData( index )
+
+		if isColor :
+
+			# in PyQt, option is passed to us correctly as a QStyleOptionViewItemV4,
+			# but in PySide it is merely a QStyleOptionViewItem and we must "cast" it.
+			if hasattr( QtGui, "QStyleOptionViewItemV4" ) :
+				option = QtGui.QStyleOptionViewItemV4( option )
+
+			transformedColor = GafferUI.DisplayTransform.get()( value )
+
+			backgroundBrush = QtGui.QBrush(
+				QtGui.QColor(
+					transformedColor.r * 255.0,
+					transformedColor.g * 255.0,
+					transformedColor.b * 255.0,
+					255.0 if isinstance( value, imath.Color3f ) else value.a * 255.0
+				)
+			)
+
+			painter.fillRect( option.rect, backgroundBrush )
+		else :
+			_Delegate.paint( self, painter, option, index )
+
+	def createEditor( self, parent, option, index ) :
+
+		w = _Delegate.createEditor( self, parent, option, index )
+
+		isColor, value = self.__colorData( index )
+		if isColor :
+			w.setParent( None )
+			# Restore popoup flag after `setParent()` above
+			w.setWindowFlags( QtCore.Qt.Popup )
+
+		return w
+
+	def _createEditorInternal( self, index ) :
+
+		isColor, value = self.__colorData( index )
+
+		if isColor :
+			self.__colorChooser = GafferUI.ColorChooser( value )
+			self.__popup = GafferUI.PopupWindow( "", child = self.__colorChooser )
+
+			# We hand over control of the widget to Qt in `createEditor()`, which
+			# will make the widget visible, but first we have to notify the Gaffer
+			# widget that it's visible to keep them in sync.
+			self.__popup.setVisible( True )
+
+			return self.__popup
+		else :
+			return GafferUI.NumericWidget( value )
+
+	def updateEditorGeometry( self, editor, option, index ) :
+
+		isColor, value = self.__colorData( index )
+		if isColor :
+			center = GafferUI.Widget.mousePosition()
+			size = editor.sizeHint()
+			topLeft = center - imath.V2i( size.width() / 2, size.height() / 2 )
+
+			editor.setGeometry( topLeft.x, topLeft.y, size.width(), size.height() )
+		else :
+			_Delegate.updateEditorGeometry( self, editor, option, index )
+
+	def setEditorData( self, editor, index ) :
+
+		isColor, value = self.__colorData( index )
+		if isColor :
+			if self.__colorChooser is not None :
+				self.__colorChooser.setColor( value )
+			else :
+				QtWidgets.QStyledItemDelegate.setEditorData( self, editor, index )
+
+		else :
+			_Delegate.setEditorData( self, editor, index )
+
+	def setModelData( self, editor, model, index ) :
+
+		if self.__colorChooser is not None :
+			model.setData( index, GafferUI._Variant.toVariant( self.__colorChooser.getColor() ), QtCore.Qt.EditRole )
+		else :
+			_Delegate.setModelData( self, editor, model, index )
+
+	# returns a tuple of the form ( isColor, colorValueOrChannelValue )
+	def __colorData( self, index ) :
+
+		data = GafferUI._Variant.fromVariant( index.model().data( index, QtCore.Qt.DisplayRole ) )
```

Comment:
Related to my previous QVariant comment, I don't think this is quite working in the way we might think it is. It's not possible for the Color3f object to be passed through the Qt C++ API (even using PySide), but here we're "skimming the surface" by accessing it from the Python override directly, so Qt never needs to see the value. If we were to call `index.data()` then we'd be going via the Qt's C++ layer, and I presume it would fail.

I'm not sure this is actually a problem, and I haven't tested my theory, but I thought it might be worth mentioning...

Reply:
Huh, it seems to work fine when I change it to `index.data()`. Perhaps that's still skirting through the Python codepath?

Reply:
Looks like PySide is way smarter than I gave it credit for!

```
> i = QtCore.QStandardItem()
> i.setData( IECoreScene.SpherePrimitive() ) # Converts to QVariant
> i.getData() # Converts back from QVariant
<IECoreScene._IECoreScene.SpherePrimitive object at 0x7f361e9487f0>
```

Sorry for the noise...